src/controlobjectscript.cpp

```diff
@@ -0,0 +1,55 @@
+#include <QApplication>
+#include <QtDebug>
+
+#include "controlobjectscript.h"
+
+ControlObjectScript::ControlObjectScript(const ConfigKey& key, QObject* pParent)
+        : ControlObjectSlave(key, pParent) {
+}
+
+void ControlObjectScript::connectScriptFunction(
+        const ControllerEngineConnection& conn) {
+    if (m_connectedScriptFunctions.isEmpty()) {
+        // we connect the slots only, if there will be actually a script
+        // connected
+        connect(m_pControl.data(), SIGNAL(valueChanged(double, QObject*)),
+                this, SLOT(slotValueChanged(double,QObject*)),
+                Qt::QueuedConnection);
+        connect(this, SIGNAL(trigger(double, QObject*)),
+                this, SLOT(slotValueChanged(double,QObject*)),
+                Qt::QueuedConnection);
+    }
+    m_connectedScriptFunctions.append(conn);
+}
+
+bool ControlObjectScript::disconnectScriptFunction(
+        const ControllerEngineConnection& conn) {
+    bool ret = m_connectedScriptFunctions.removeAll(conn) > 0;
+    if (m_connectedScriptFunctions.isEmpty()) {
+        // no script left, we can disconnected
+        disconnect(m_pControl.data(), SIGNAL(valueChanged(double, QObject*)),
+                this, SLOT(slotValueChanged(double,QObject*)));
+        disconnect(this, SIGNAL(trigger(double, QObject*)),
+                this, SLOT(slotValueChanged(double,QObject*)));
+    }
+    return ret;
+}
+
+void ControlObjectScript::slotValueChanged(double value, QObject*) {
+    // Make a local copy of m_connectedScriptFunctions fist.
+    // This allows a script to disconnect a callback from the callback
+    // itself. Otherwise the this may crash since the disconnect call
+    // happens during conn.function.call() in the middle of the loop below.
+    QList<ControllerEngineConnection> connections = m_connectedScriptFunctions;
+    for(auto&& conn: connections) {
```

Comment:
Cool, I think I have learned a bit. 

> However, it won't compile for proxy objects - for example, if the
range is vector<bool>, then *__begin will be a temporary object returned by
value (of type vector<bool>::reference), and auto& refuses to bind to
temporaries.

bool is std::_Bit_reference in this case. 

This means "auto&&" will always work "auto&" only works if the container and the stored type allows it. In our case it works and it is probably less confusing for a reader. 

If a "auto&&" is an rvalue reference, is it possible to write the value back? 
Interestingly you can modify the value, but it is not persistent since the evaluated rvalue exists only in a register or so. 

int&& rvalue = 3 * 2;
int& lvalue = rvalue; 
lvalue = 4; 
pintf("%d %d", lvalue, rvalue); 

output: 
4 4 

However why does this work? 

        std::vector<bool> hallo = { false, true, false };

        for(auto&& element: hallo) {
                element = false;
                printf("value %d\n", (int)element);
        }

        for(const auto& element: hallo) {
                printf("value %d\n", (int)element);
        }

output 
0
0
0
0
0